docker/Dockerfile.build

```diff
@@ -1,56 +1,87 @@
-# Builder image
-FROM rust:buster as builder
+# This steps implement cargo-chef for docker layer caching
+# We are using four stages:
+# - chef: installs cargo-chef
+# - planner: computes the recipe file
+# - builder: caches our dependencies and builds the binary
+# - tester: builds and run tests
+# - runtime: is our runtime environment
+FROM rust:bullseye as chef
+RUN cargo install cargo-chef --locked
+WORKDIR /app
+
+FROM chef AS planner
+COPY . .
+RUN cargo chef prepare --recipe-path recipe.json
+
+FROM chef AS builder
+COPY --from=planner /app/recipe.json recipe.json
 
 RUN apt-get update && \
-	apt-get install -y --no-install-recommends \
-	make cmake g++ gcc llvm libclang-dev clang ca-certificates
+    apt-get install -y --no-install-recommends \
+    llvm libclang-dev clang ca-certificates && \
+    rm -rf /var/lib/apt/lists/* /tmp/*
 
-RUN mkdir /zebra
-WORKDIR /zebra
+# Optimize builds. In particular, regenerate-stateful-test-disks.yml was reaching the
+# GitHub Actions time limit (6 hours), so we needed to make it faster.
+ENV RUSTFLAGS -O
+ENV CARGO_HOME /app/.cargo/
+# Build dependencies - this is the caching Docker layer!
+RUN cargo chef cook --release --features enable-sentry --recipe-path recipe.json
 
-ARG SHORT_SHA
-ENV SHORT_SHA $SHORT_SHA
+# Build zebra
+ARG RUST_BACKTRACE=1
+ENV RUST_BACKTRACE ${RUST_BACKTRACE}
 
-ENV RUST_BACKTRACE full
-ENV CARGO_HOME /zebra/.cargo/
+COPY . .
+RUN cargo build --locked --jobs 2 --release --features enable-sentry --bin zebrad
 
-RUN rustc -V; cargo -V; rustup -V
+FROM builder AS tester
+# Skip IPv6 tests by default, as some CI environment don't have IPv6 available
+ARG ZEBRA_SKIP_IPV6_TESTS=1
+ENV ZEBRA_SKIP_IPV6_TESTS ${ZEBRA_SKIP_IPV6_TESTS}
 
-COPY . .
+RUN cargo test --locked --jobs 2 --release --features enable-sentry --workspace --no-run
+RUN cargo run --bin zebrad download
 
-RUN cd zebrad/; cargo build --release --features enable-sentry
+CMD ["cargo" "test" "--workspace"]
```

Comment:
I think these will trigger an entire new build since they don't have `--release`

Reply:
I meant to `run` the compiled/available test, not to re-`compile` them. How can I accomplish that?

Reply:
You need to match the `cargo test` line above, e.g. add `--locked --release --features enable-sentry`